internal/driver/basenotificationconsumer.go

```diff
@@ -0,0 +1,152 @@
+	consumer.lc.Infof("Consumer starts the Renew loop for '%s'", consumer.Name)
+	// Remove self when subscription finished or renew failed
+	defer consumer.manager.removeConsumer(consumer)
+
+	duration, err := ParseISO8601(*consumer.subscriptionRequest.InitialTerminationTime)
+	if err != nil {
+		consumer.lc.Infof("invalid Initial termination time, %v", err)
+		return
+	}
+
+	// Send Renew request every ten second before termination time
+	renewTime := duration - 10*time.Second
+	renewTicker := time.NewTicker(renewTime)
+	for {
+		select {
+		case <-consumer.Stopped:
+			consumer.lc.Infof("Finish the subscription '%s'", consumer.Name)
+			return
+		case <-renewTicker.C:
+			consumer.lc.Debugf("Renew the subscription from '%s' for resource '%s'", consumer.SubscriptionAddress, consumer.Name)
+			renewRequest := consumer.createRewRequest()
+			renewRequestData, err := xml.Marshal(renewRequest)
+			if err != nil {
+				consumer.lc.Errorf("Fail to marshal subscription request for resource '%s'", consumer.Name)
+				return
+			}
+
+			servResp, err := consumer.deviceClient.onvifDevice.SendSoap(consumer.SubscriptionAddress, string(renewRequestData))
+			if err != nil {
+				consumer.lc.Warnf("Fail to send the renew request from '%s' for resource '%s', %v. The pull point expired or dropped, try to create a new one.", consumer.SubscriptionAddress, consumer.Name, err)
+				err = consumer.subscribe()
+				if err != nil {
+					consumer.lc.Errorf("Fail to subscribe again for resource '%s', %v", consumer.Name, err)
+					return
+				}
+			} else if servResp.StatusCode >= http.StatusBadRequest {
+				response, err := renewResponse(servResp)
+				consumer.lc.Warnf("Fail to renew the subscription from '%s' for resource '%s', status code: %s, err: %v. The pull point expired or dropped, try to create a new one.", consumer.SubscriptionAddress, consumer.Name, response.Body.Fault.String())
+				err = consumer.subscribe()
+				if err != nil {
+					consumer.lc.Errorf("Fail to subscribe again for resource '%s', %v", consumer.Name, err)
+					return
+				}
+			}
+		}
+	}
+}
+
+func (consumer *Consumer) subscribe() errors.EdgeX {
+	subscribe := consumer.subscribeRequest()
+	subscribeData, err := json.Marshal(subscribe)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("fail to marshal subscription request for resource '%s'", consumer.Name), err)
+	}
+	serviceName := onvif.EventWebService
+	functionName := onvif.Subscribe
+	respContent, edgexErr := consumer.deviceClient.callOnvifFunction(serviceName, functionName, subscribeData)
+	if edgexErr != nil {
+		return errors.NewCommonEdgeX(errors.Kind(edgexErr), fmt.Sprintf("fail to subscribe again for resource '%s', %v", consumer.Name, err), edgexErr)
+	}
+	subscribeResponse := respContent.(*event.SubscribeResponse)
+	consumer.SubscriptionAddress = fmt.Sprint(subscribeResponse.SubscriptionReference.Address)
+	return nil
+}
+
+func (consumer *Consumer) createRewRequest() *event.Renew {
```

Comment:
should be? or whate is `rew`
```suggestion
func (consumer *Consumer) createRawRequest() *event.Renew {
```

Reply:
Renamed the func.